ci/gpu/build.sh

```diff
@@ -61,6 +61,9 @@ logger "Build and install librmm and rmm..."
 if hasArg --skip-tests; then
     logger "Skipping Tests..."
 else
+    logger "Installing extra test dependencies..."
```

Comment:
Is this the way we normally handle python test dependencies? Why not just add it to the conda env recipe?

Reply:
That is what we do in ucx-py but we could be doing it wrong :)
Is it possible to add it to the conda environment recipe without having the released rmm conda package depend on cupy?

Reply:
We should update the Dockerfiles here if possible: https://github.com/rapidsai/gpuci-build-environment so that we're not constantly dependent on conda downloads for the dependencies

Reply:
What about for developers who don't use Docker?

Reply:
Users can still install `cupy` with Conda. Shouldn't be an issue (unless I'm missing something).

Reply:
My understanding of Keith's statement (though please feel free to correct me Keith) is we are starting rely on CuPy in a lot of places: RMM, cuDF, cuML, ucx-py, and maybe more. So for our development process it might just make sense to have CuPy around in the image.

The flipside of that is we may be unintentionally adding CuPy to other libraries as it would seep in due to it already being in the CI environment and so we might assume it is there without explicitly requiring it.

We can balance this by caching the CuPy package in the image, but not installing it anywhere. If we did this, we would keep these lines (or equivalent lines based on OPS direction) and we would add a line to the image to download the CuPy package only.

Thoughts on this?

Reply:
I don't have a strong opinion. It just seems weird to me to silently install a dependency *during testing*.

Reply:
I mean, if I'm a developer and I use build.sh to run the tests (I don't do this) it might be surprising that running the tests modified my conda environment.

Reply:
I see. Are you running `ci/gpu/build.sh` to test locally? Is this how others are testing RMM? If so, I can see how this might be a problem.

Reply:
I don't use build.sh at all, currently.  I rely on the libcudf conda env recipe including all dependencies I need.

Reply:
Yeah if you install `cudf`, `cupy` is a hard requirement there. So you will get it. If you are just using the C++ side, `cupy` won't be relevant.

FWIW I asked OPS on Slack if they could weigh in here (in case we've missed something). Hopefully they can point us in the right direction when they have a moment.

Reply:
@kkraus14 is right. If it's something that's going to be used in multiple projects or maybe is just a large install, it can go in the gpuCI build environment. If it's a one off or very specific, doing a conda install in this script is okay. Overall though, I'd avoid conda installing in this script as much as possible as it can complicate debugging and may have undesirable side effects.

As to @harrism point, there should probably be a separate conda environment file to list out the test dependencies for those who don't rely on this script.

Reply:
Thanks Ray! 😄

Ok, so if we are to add this to gpuCI, where would we add it? Would [here]( https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile.centos7 ) be appropriate or should it go somewhere else?

That makes sense. So where should we add this? Would [these environment files]( https://github.com/rapidsai/rmm/tree/branch-0.12/conda/environments ) be the right place? Or should it go somewhere else?

Reply:
@raydouglass @kkraus14 do you want this dependency issue to be handled in this PR like @jakirkham suggest? Or should it go in a following PR?

Reply:
> Thanks Ray! 😄
> 
> Ok, so if we are to add this to gpuCI, where would we add it? Would [here](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile.centos7) be appropriate or should it go somewhere else?

Both [Dockerfile](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile) and [Dockerfile.centos7](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile.centos7)

> That makes sense. So where should we add this? Would [these environment files](https://github.com/rapidsai/rmm/tree/branch-0.12/conda/environments) be the right place? Or should it go somewhere else?

Yes

> @raydouglass @kkraus14 do you want this dependency issue to be handled in this PR like @jakirkham suggest? Or should it go in a following PR?

If the dependency is required for this PR and it's going to be installed in `ci/gpu/build.sh` then it needs to be in this PR. If it's going in gpuCI images, then this PR needs to wait for those images to be updated. You can open a PR in https://github.com/rapidsai/gpuci-build-environment for that.

Reply:
> > Ok, so if we are to add this to gpuCI, where would we add it? Would [here](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile.centos7) be appropriate or should it go somewhere else?
> 
> Both [Dockerfile](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile) and [Dockerfile.centos7](https://github.com/rapidsai/gpuci-build-environment/blob/master/Dockerfile.centos7)

I went ahead and added PR ( https://github.com/rapidsai/gpuci-build-environment/pull/55 ) to add CuPy to the Docker images.

> > That makes sense. So where should we add this? Would [these environment files](https://github.com/rapidsai/rmm/tree/branch-0.12/conda/environments) be the right place? Or should it go somewhere else?
> 
> Yes

@madsbk, would you be comfortable doing this?